File: rust/src/questions/k05.rs
use crate::shared::structs::golem::Golem;
use crate::shared::traits::executable::Executable;
use crate::shared::utility::read_user_input;
use rand::prelude::*;
use std::io::Write;

pub struct K05 {}

impl Executable for K05 {
    fn question_1(&self) -> anyhow::Result<()> {
        let mut salary: f32 = 19.0;
        let mut age: i32 = 22;

        while salary < 50.0 {
            salary *= 1.035;
            age += 1;
        }

        println!("{}歳で月給{}万円", age, salary);
        Ok(())
    }

    fn question_2(&self) -> anyhow::Result<()> {
        let mut choice = 0;

        while choice != 1 {
            println!("起きろ～");
            print!("1．起きた　2．あと5分…　3．Zzzz…\t入力：");
            std::io::stdout().flush()?;
            choice = read_user_input::<i32>()?;
        }
        println!("よし、学校へ行こう！");
        Ok(())
    }

    fn question_3(&self) -> anyhow::Result<()> {
        loop {
            println!("起きろ～");
            print!("1．起きた　2．あと5分…　3．Zzzz…\t入力：");
            std::io::stdout().flush()?;
            let choice = read_user_input::<i32>()?;
            if choice == 1 {
                println!("よし、学校へ行こう！");
            }
        }
    }

    fn question_4(&self) -> anyhow::Result<()> {
        let rand_num = rand::thread_rng().gen_range(0..201);
        let mut golem = Golem::new(300 + rand_num);
        let mut player_hp = 200 + rand::thread_rng().gen_range(0..101);

        println!("ゴーレム　（HP:{}　防御力：{}）", golem.hp, golem.defense);

        while golem.hp > 0 {
            println!("HP：{}", golem.hp);
            print!("攻撃手段を選択してください（1．攻撃　2．特技　3．魔法）　＞");
            std::io::stdout().flush()?;

            let choice = read_user_input::<i32>()?;

            let damage = match choice {
                1 => 60 + rand::thread_rng().gen_range(0..41),
                2 => 30 + rand::thread_rng().gen_range(0..101),
                3 => 20 + rand::thread_rng().gen_range(0..181),
                _ => 0,
            };

            println!("基礎攻撃力は{}です。", damage);
            let damage = if damage - golem.defense <= 0 {
                0
            } else {
                damage - golem.defense
            };
            if damage == 0 {
                println!("ゴーレム：「ハハハハハ、情けないな！貴様は弱すぎる！」");
                println!("ゴーレムがあなたを攻撃しました！攻撃値：{}", golem.attack);
                player_hp -= golem.attack;

                if player_hp <= 0 {
                    println!("あなたはゴーレムに負けました！ゲームオーバー！");
                    return Ok(());
                }
                println!("あなたの残りHPは：{}", player_hp);
            }
            println!("ダメージは{}です。", damage);
            golem.hp -= damage;

            if golem.hp < 0 {
                golem.hp = 0;
            }

            println!("残りのHPは{}です。", golem.hp);
            if golem.hp == 0 {
                println!("ゴーレムを倒しました！");
            }
        }
        Ok(())
    }
}

File: TypeScript/assignments/K05.ts
import Executable from "../shared/Executable.ts"
import Golem from "../shared/Golem.ts";
import { getRandomNumber } from "../shared/Utility.ts";

export default class K05 extends Executable {
    question1(): void {
        let salary = 19.0;
        let age = 22;

        while (salary < 50.0) {
            salary *= 1.035;
            age++;
        }

        console.log(`${age}歳で月給${salary}万円`);
    }

    question2(): void {
        let choice = 0;
        do {
            console.log("起きろ～");
            choice = Number.parseInt(prompt("1．起きた　2．あと5分…　3．Zzzz…  入力：") ?? "0");
        } while (choice != 1);
        console.log("よし、学校へ行こう！");
    }

    question3(): void {
        do {
            console.log("起きろ～");
            const choice = Number.parseInt(prompt("1．起きた　2．あと5分…　3．Zzzz…  入力：") ?? "0");
            if (choice === 1) {
                console.log("よし、学校へ行こう！");
            }
        } while (true);
    }

    question4(): void {
        const golem: Golem = {
            hp: 300 + getRandomNumber(201),
            defense: 80,
            attack: 50
        };

        let playerHp = 200 + getRandomNumber(101);

        console.log(`ゴーレム　（HP：${golem.hp}　防御力：${golem.defense}）`);

        while (golem.hp > 0) {
            console.log("残りHP：" + golem.hp);
            const choice = Number.parseInt(prompt("攻撃手段を選択してください（1．攻撃　2．特技　3．魔法）　＞") ?? "0");

            const damageFunc = (c: number) => {
                switch (c) {
                    case 1:
                        return 60 + getRandomNumber(41);
                    case 2:
                        return 30 + getRandomNumber(101);
                    case 3:
                        return 20 + getRandomNumber(181);
                    default:
                        return 0;
                }
            };

            let damage = damageFunc(choice);
            console.log(`基礎攻撃力は${damage}です。`);
            damage = damage - golem.defense < 0 ? 0 : damage - golem.defense;

            if (damage === 0) {
                console.log("ゴーレム：「ハハハハハ、情けないな！貴様は弱すぎる！」");
                console.log("ゴーレムがあなたを攻撃しました！攻撃値：" + golem.attack);
                playerHp -= golem.attack;

                if (playerHp <= 0) {
                    console.log("あなたはゴーレムに負けました！ゲームオーバー！");
                    return;
                }
                console.log("あなたの残りHPは：" + playerHp);
            }

            console.log(`ダメージは${damage}です`);
            golem.hp = golem.hp - damage < 0 ? 0 : golem.hp - damage;
            console.log(`残りのHPは${golem.hp}です`);

            if (golem.hp === 0) {
                console.log("ゴーレムを倒しました！");
            }
        }
    }
}